File: applications/go-service/tracing.go
package main

import (
	"context"
	"os"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	"go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
)

func initTracing() (*trace.TracerProvider, error) {
	ctx := context.Background()

	// Get Jaeger collector URL from environment
	// Jaeger accepts OTLP HTTP on port 4318 (default) or 14268 (legacy HTTP)
	jaegerHost := os.Getenv("JAEGER_COLLECTOR_HOST")
	if jaegerHost == "" {
		jaegerHost = "jaeger-query.monitoring.svc.cluster.local"
	}

	// Create OTLP HTTP exporter pointing to Jaeger collector
	exporter, err := otlptracehttp.New(ctx,
		otlptracehttp.WithEndpoint("http://"+jaegerHost+":4318/v1/traces"),
		otlptracehttp.WithInsecure(),
	)
	if err != nil {
		return nil, err
	}

	// Create resource with service information
	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName("go-service"),
			semconv.ServiceVersion("1.0.0"),
		),
	)
	if err != nil {
		return nil, err
	}

	// Create trace provider
	tp := trace.NewTracerProvider(
		trace.WithBatcher(exporter),
		trace.WithResource(res),
		trace.WithSampler(trace.AlwaysSample()),
	)

	// Set global tracer provider
	otel.SetTracerProvider(tp)

	// Set global propagator
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	return tp, nil
}

File: applications/python-service/tracing.py
#!/usr/bin/env python3
"""
OpenTelemetry tracing initialization for Python service
"""

import os
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.semantic_conventions.resource import ResourceAttributes

def init_tracing():
    """Initialize OpenTelemetry tracing"""
    # Get Jaeger collector URL from environment
    jaeger_host = os.getenv("JAEGER_COLLECTOR_HOST", "jaeger-query.monitoring.svc.cluster.local")
    jaeger_url = f"http://{jaeger_host}:4318/v1/traces"
    
    # Create resource with service information
    resource = Resource.create({
        ResourceAttributes.SERVICE_NAME: "python-service",
        ResourceAttributes.SERVICE_VERSION: "1.0.0",
    })
    
    # Create tracer provider
    tracer_provider = TracerProvider(resource=resource)
    
    # Create OTLP HTTP exporter
    otlp_exporter = OTLPSpanExporter(
        endpoint=jaeger_url,
        insecure=True,
    )
    
    # Add span processor
    span_processor = BatchSpanProcessor(otlp_exporter)
    tracer_provider.add_span_processor(span_processor)
    
    # Set global tracer provider
    trace.set_tracer_provider(tracer_provider)
    
    print(f"OpenTelemetry tracing initialized for python-service (Jaeger: {jaeger_url})")
    
    return tracer_provider

